blog/2021-12-rxjs-turn-signal/README.md

```diff
@@ -0,0 +1,312 @@
+```ts
+const closingNotifier$ = forkJoin([
+  release$.pipe(take(1)),
+  timer(stepTimeMs * cycles * 2),
+]);
+```
+
+### Putting things together
+
+Now that we have created the notifier that tells us when the signalling period has to be over, we can put things together.
+What we want to terminate is the individual signal stream that the press event is mapped to.
+Right now, we convert each trigger press event to to an endlessly emitting signal event stream.
+This is the part we have to change: We want the signalling to eventually come to an end.
+It is therefore important to apply the `takeUntil()` operator to the inner mapped Observable, not to the outer stream that is derived from `press$`.
+
+```ts
+press$.pipe(
+  exhaustMap(() => rawSignal$.pipe(
+      takeUntil(closingNotifier$)
+    )
+  )
+)
+```
+
+## Switching off the light
+
+We can now see that the signal comes to an end after at least 3 cycles and when we release the button.
+However, there's an tiny edge case:
+When we press the button for longer than 3 cycles and then release it while the signal is active, the light stays on!
+We have interrupted the signal stream somewhere in the middle and no subsequent emission made the light go off.
+
+To change this, we have to tell our signal stream to always end with going off.
+So whenever the inner mapped Observable completes, it should emit `false`.
+Say hello to the `endWith()` operator!
+
+```ts
+press$.pipe(
+  exhaustMap(() => rawSignal$.pipe(
+      takeUntil(closingNotifier$),
+      endWith(false)
+    )
+  )
+)
+```
+
+Try it out: Press the trigger button for a longer time. Then release it while the light is on. It will immediately be switched off.
+
+## Bonus: Making it mobile-friendly
+
+In the beginning, we used `mousedown` and `mouseup` events to capture trigger press and release.
+This works fine in a desktop browser. However, on mobile touch devices we rather have to use `touchstart` and `touchend` events instead.
+
+Since the browser only emits one or the other event, we can combine them and use the first one that is available.
+This can be achieved with the `race()` creation function: It listens to multiple Observables at once. When one of them emits, it wins the race and all of the others are unsubscribed.
+In our case, whoever comes first – `mousedown` oder `touchstart` – will be used.
+
+```ts
+const press$ = race(
+  fromEvent(signalEl, 'mousedown'),
+  fromEvent(signalEl, 'touchstart')
+);
+const release$ = race(
+  fromEvent(document, 'mouseup'),
+  fromEvent(document, 'touchend')
+);
+```
+
+
+
+## Wrap up
+
+With relatively few lines of code we have created a car turn indicator with RxJS!
+Each trigger press is converted to an alternating signal.
+While a signal cycle is running, all other presses will be ignored.
+The signal ends after the trigger is released and at least 3 cycles have been finished.
+Regardless of how the interrupted signal ends, it will always switch off the light when it leaves the room.
+
+
```

Comment:
Hier bitte nochmal der volle Quelltext rein. Wenn der Stackblitz link mal irgendwann nicht mehr geht, dann ist die Info weg. Und nur Github-Zeugs kommt in den GitHub Arctic Code Vault – denk an die Menschheit in 10.000 Jahren! 😉